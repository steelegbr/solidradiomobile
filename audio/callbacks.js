/**
 * Audio player callbacks.
 */

 module.exports = async function() {
     
 }