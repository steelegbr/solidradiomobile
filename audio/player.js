/**
 * Indicates the possible player states.
 */

export const PlayerState = {
    UNINITIALISED: 0,
    IDLE: 1,
    LOADING: 3,
    PLAYING: 4,
    PAUSED: 5,
    ERROR: 6
}