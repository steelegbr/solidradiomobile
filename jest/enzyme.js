/**
    Solid Radio Mobile App
    Copyright (C) 2020-2021 Marc Steele

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import 'react-native';
import 'jest-enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';

/**
 * Set up Enzyme to mount to DOM, simulate events,
 * and inspect the DOM in tests.
 */
Enzyme.configure({ adapter: new Adapter() });

// JSDOM - required for Enzyme's mount

const { JSDOM } = require('jsdom');
const jsdom = new JSDOM('<!doctype html><html><body></body></html>');
const { window } = jsdom;
global.window = window;
global.document = window.document;