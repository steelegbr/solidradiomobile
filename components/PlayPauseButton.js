import React, { Component } from 'react';
import { ActivityIndicator } from 'react-native-paper';
import { connect } from 'react-redux';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { PlayerState } from '../audio/player';
import { togglePlayPause } from '../reducers/actions';

class PlayPauseButton extends Component {

    render() {

        const { icon, style, bigMode, iconSize, showSpinner, togglePlayPause } = this.props;

        if (showSpinner) {

            // Loading spinner

            var size = bigMode ? iconSize : 16;
            return(
                <ActivityIndicator size={size} />
            );

        } else {

            // Normal button with play/pause options

            if (bigMode) {
                return(
                    <Icon.Button name={icon} size={iconSize} color={style.color} backgroundColor={style.backgroundColor} onPress={togglePlayPause} />
                );
            } else {
                return(
                    <Icon.Button name={icon} size={16} style={style} color={style.color} backgroundColor={style.backgroundColor} onPress={togglePlayPause} />
                );
            }

        }

    }

}

function mapStateToProps(state, ownProps) {

    // Size overrides

    const bigMode = 'bigMode' in ownProps ? ownProps.bigMode : false;
    const iconSize = 'iconSize' in ownProps ? ownProps.iconSize : 0;

    // Set the icon based on the state we're in

    let icon;
    let showSpinner = false;

    switch (state.player.state) {
        case PlayerState.UNINITIALISED:
        case PlayerState.IDLE:
        case PlayerState.ERROR:
        case PlayerState.PAUSED:
            icon = "play-circle";
            break;
        case PlayerState.LOADING:
            showSpinner = true;
            break;
        case PlayerState.PLAYING:
            icon = "pause-circle";
            break;
    }

    return {
        icon: icon,
        style: {
            color: state.theme.colors.primary,
            backgroundColor: 'rgba(0, 0, 0, 0.0)',
            padding: 10,
            margin: 5
        },
        bigMode: bigMode,
        iconSize: iconSize,
        showSpinner: showSpinner
    };

}

const mapDispatchToProps = dispatch => {
    return {
        togglePlayPause: () => dispatch(togglePlayPause('play_pause_button'))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(PlayPauseButton);