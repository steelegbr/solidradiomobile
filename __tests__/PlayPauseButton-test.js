/**
    Solid Radio Mobile App
    Copyright (C) 2020-2021 Marc Steele

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

/**
 * Tests the play/pause button.
 */

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import PlayPauseButton from '../components/PlayPauseButton';
import { PlayerState } from '../audio/player';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { mount } from 'enzyme';

// Mocks

jest.mock("react-native-paper", () => ({
    ActivityIndicator: "ActivityIndicator",
    DefaultTheme: {
        colors: {

        }
    }
}));

// Redux

const mockStore = configureStore([]);

// Tests

it('renders-unitialised', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.UNINITIALISED
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-idle', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.IDLE
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-error', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.ERROR
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-paused', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.PAUSED
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-playing', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.PLAYING
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-playing-big-mode', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.PLAYING
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton bigMode={true} iconSize={40} />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-loading', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.LOADING
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('renders-loading-big-mode', () => {

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.LOADING
        }
    });

    const playPauseButton = renderer.create(
        <Provider store={store}>
            <PlayPauseButton bigMode={true} iconSize={40} />
        </Provider>
    ).toJSON();

    expect(playPauseButton).toMatchSnapshot();

});

it('play-pause-pressed', () => {

    // Arrange

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.PLAYING
        }
    });

    const playPauseButton = mount(
        <Provider store={store}>
            <PlayPauseButton />
        </Provider>
    );

    // Act

    playPauseButton.find(Icon.Button).first().props().onPress();

    // Assert

    const expectedActions = [{ "type": "AUDIO_PLAYER_PLAYPAUSE", "source": "play_pause_button" }];
    expect(store.getActions()).toStrictEqual(expectedActions);

});

it('play-pause-presse-big-mode', () => {

    // Arrange

    const store = mockStore({
        theme: {
            colors: {
                primary: "#ff0000"
            }
        },
        player: {
            state: PlayerState.PLAYING
        }
    });

    const playPauseButton = mount(
        <Provider store={store}>
            <PlayPauseButton bigMode={true} iconSize={40} />
        </Provider>
    );

    // Act

    playPauseButton.find(Icon.Button).first().props().onPress();

    // Assert

    const expectedActions = [{ "type": "AUDIO_PLAYER_PLAYPAUSE", "source": "play_pause_button" }];
    expect(store.getActions()).toStrictEqual(expectedActions);

});