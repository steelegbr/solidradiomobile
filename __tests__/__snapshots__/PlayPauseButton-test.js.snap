// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders-error 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.0)",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "borderRadius": 5,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "color": "#ff0000",
          "margin": 5,
          "padding": 10,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ff0000",
            "fontSize": 16,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders-idle 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.0)",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "borderRadius": 5,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "color": "#ff0000",
          "margin": 5,
          "padding": 10,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ff0000",
            "fontSize": 16,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders-loading 1`] = `
<ActivityIndicator
  size={16}
/>
`;

exports[`renders-loading-big-mode 1`] = `
<ActivityIndicator
  size={40}
/>
`;

exports[`renders-paused 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.0)",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "borderRadius": 5,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "color": "#ff0000",
          "margin": 5,
          "padding": 10,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ff0000",
            "fontSize": 16,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders-playing 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.0)",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "borderRadius": 5,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "color": "#ff0000",
          "margin": 5,
          "padding": 10,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ff0000",
            "fontSize": 16,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders-playing-big-mode 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.0)",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "borderRadius": 5,
        },
        undefined,
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ff0000",
            "fontSize": 40,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`renders-unitialised 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.0)",
        "borderRadius": 5,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "padding": 8,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "borderRadius": 5,
        },
        Object {
          "backgroundColor": "rgba(0, 0, 0, 0.0)",
          "color": "#ff0000",
          "margin": 5,
          "padding": 10,
        },
      ]
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#ff0000",
            "fontSize": 16,
          },
          Object {
            "marginRight": 10,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
